### PR TITLE
OMPI/COLL: coll ucc component

### PR DESCRIPTION
**Pull request by vspetrov**
Adds support for UCC (unified collective communication) library: https://github.com/openucx/ucc

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e781d9a030376d2e70e554ef207867b1

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5eedc11e8cb8c375fedb29b956f7d3ee

**Comment by jjhursey** [comment]
IBM CI had a bad weekend (sorry about that). We should be all good now.
bot:ibm:pgi:retest
bot:ibm:xl:retest

**Comment by vspetrov** [comment]
@jladd-mlnx @bureddy @Sergei-Lebedev @manjugv  i can't set reviewers here, plz start looking

File: ompi/mca/coll/ucc/coll_ucc_module.c [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+static int mca_coll_ucc_module_enable(mca_coll_base_module_t *module,
+                                      struct ompi_communicator_t *comm)
+{
+    mca_coll_ucc_component_t *cm         = &mca_coll_ucc_component;
+    mca_coll_ucc_module_t    *ucc_module = (mca_coll_ucc_module_t *)module;
+    ucc_status_t              status;
+    int rc;
+    ucc_team_params_t team_params = {
+        .mask               = UCC_TEAM_PARAM_FIELD_EP |
+                              UCC_TEAM_PARAM_FIELD_EP_RANGE |
```
Comment (Sergei-Lebedev) [inline ompi/mca/coll/ucc/coll_ucc_module.c:292]: ep_range field is not set

File: ompi/mca/coll/ucc/coll_ucc_ops.c [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+int mca_coll_ucc_barrier(struct ompi_communicator_t *comm,
+                         mca_coll_base_module_t *module)
+{
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_coll_req_h         req;
+
+    UCC_VERBOSE(3, "running ucc barrier");
+
+    ucc_coll_args_t coll = {
+        .coll_type = UCC_COLL_TYPE_BARRIER
```
Comment (Sergei-Lebedev) [inline ompi/mca/coll/ucc/coll_ucc_ops.c:45]: i think it's better to set mask to zero here

File: ompi/mca/coll/ucc/coll_ucc_ops.c [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_datatype_t         ucc_dt     = ompi_dtype_to_ucc_dtype(dtype);
+    ucc_coll_req_h         req;
+    UCC_VERBOSE(3, "running ucc bcast");
+    if (COLL_UCC_DT_UNSUPPORTED == ucc_dt) {
+        goto fallback;
+    }
+
+    ucc_coll_args_t coll = {
+        .coll_type = UCC_COLL_TYPE_BCAST,
```
Comment (Sergei-Lebedev) [inline ompi/mca/coll/ucc/coll_ucc_ops.c:70]: set mask 0?

File: ompi/mca/coll/ucc/coll_ucc_ops.c [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    ucc_sdt = ompi_dtype_to_ucc_dtype(sdtype);
+    ucc_rdt = ompi_dtype_to_ucc_dtype(rdtype);
+    if (COLL_UCC_DT_UNSUPPORTED == ucc_sdt ||
+        COLL_UCC_DT_UNSUPPORTED == ucc_rdt) {
+        goto fallback;
+    }
+
+    opal_datatype_type_size(&sdtype->super, &dt_size);
+    ucc_coll_args_t coll = {
```
Comment (Sergei-Lebedev) [inline ompi/mca/coll/ucc/coll_ucc_ops.c:163]: here we definitely need to set mask to zero. If it contains garbage we can end up with inplace alltoall accidently

File: ompi/mca/coll/ucc/coll_ucc.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    mca_coll_base_module_allreduce_fn_t previous_allreduce;
+    mca_coll_base_module_t*             previous_allreduce_module;
+    mca_coll_base_module_barrier_fn_t   previous_barrier;
+    mca_coll_base_module_t*             previous_barrier_module;
+    mca_coll_base_module_bcast_fn_t     previous_bcast;
+    mca_coll_base_module_t*             previous_bcast_module;
+    mca_coll_base_module_alltoall_fn_t  previous_alltoall;
+    mca_coll_base_module_t*             previous_alltoall_module;
+    mca_coll_base_module_alltoallv_fn_t previous_alltoallv;
+    mca_coll_base_module_t*             previous_alltoallv_module;
```
Comment (bureddy) [inline ompi/mca/coll/ucc/coll_ucc.h:70]: how about non-blocking collectives?

Comment (vspetrov) [reply ompi/mca/coll/ucc/coll_ucc.h]: updated the PR to support NB collectives using UCC API completion callback. plz review.

**Comment by vspetrov** [comment]
@jsquyres @jladd-mlnx @manjugv i don't have permissions to set Milestone to "Target: master" and "Target: v5.0.x". Neither can I set reviewers. Please assist.

**Comment by jsquyres** [comment]
@vspetrov I think I just gave you the appropriate access.

File: ompi/mca/coll/ucc/coll_ucc.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+#include "coll_ucc_debug.h"
+#include <ucc/api/ucc.h>
+
+BEGIN_C_DECLS
+
+#define COLL_UCC_CTS (UCC_COLL_TYPE_BARRIER   | UCC_COLL_TYPE_BCAST    | \
+                      UCC_COLL_TYPE_ALLREDUCE | UCC_COLL_TYPE_ALLTOALL | \
+                      UCC_COLL_TYPE_ALLTOALLV)
+
+#define COLL_UCC_CTS_STR "barrier,bcast,allreduce,alltoall,alltoallv"
```
Comment (bureddy) [inline ompi/mca/coll/ucc/coll_ucc.h:31]: do we need to add non-blocking colls in this list? (or) name here represents both

Comment (vspetrov) [reply ompi/mca/coll/ucc/coll_ucc.h:31]: i will add separate names i think - lets be explicit.

**Review by janjust — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Review by wfaderhold21 — APPROVED** [review]

**Review by bureddy — APPROVED** [review]

**Pull request merged**